### PR TITLE
Ubuntu Copy-From-Host collectors

### PR DESCRIPTION
**Pull request by danbudris**
https://github.com/aws/eks-anywhere/issues/238

*Description of changes:*

This adds support for copy-from-host functionality to the support bundle.

This allows us to actually copy logs from the underlying worker nodes, archive them, and return them in the support bundle.

For example, `eksctl anywhere generate support-bundle -f MyClusterConfig.yaml` will ingest the cluster configuration, generate the appropriate collectors based on the OS family of the machine configs, launch a pod on the cluster based on our collector image from the bundle mapping the host volumes, and then copy logs from the volumes on the worker nodes of the cluster to an archive.

We're currently collecting the cloud-init logs as well as the `syslog`, which contains the `kubelet` and `containerd` logs, among others. This will expand over time. I'm going to take it as a task to better educate myself on the logging architecture so we can determine if there's a better way to be outputting and ingesting the `systemd` logs...it's too bad we can't access `journalctl` on the hosts.

A change to note  -- I've added an `OSFamily()` method to the `MachineConfigs` provider interface. This allows us to determine what OS family the machines are, regardless of the provider type; an alternative solution would be to switch on the datacenter config Kind and cast the machine configs accordingly, to extract the OS family. After conversations with @jiayiwang7 it seems like this is the better approach, even if we have to instantiate the relatively heavy-weight provider in order to do so, as the logic about the provider spec remains encapsulated.

** Testing **
created, upgraded, and delete clusters
collected support bundles, generated support bundles

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/support/runsupportbundle.go
Comment (vivek-koppuru) [inline pkg/support/runsupportbundle.go:75]: Nit: The `.` should go in the front instead of in the end right?

File: pkg/support/collectors.go
Comment (vivek-koppuru) [inline pkg/support/collectors.go:121]: Instead of making it a switch here, can we just have a map of osFamily to the appropriate method? That way instead of having to loop through all the machine config objects at the top to create our set and then also setting it to true, we can just retrieve the appropriate collectors based on the machine config. Maybe I am missing some other reason as to why we are doing it this way though.

File: test/framework/flux.go
Comment (vivek-koppuru) [inline test/framework/flux.go:234]: Are you making this change because you saw an error in dereferencing? Wondering if we want to add a check for nil if that's the case.

File: pkg/providers/provider.go
Comment (vivek-koppuru) [inline pkg/providers/provider.go:51]: This interface was mainly created along with the `MachineConfigs` method to make it easier to output the yaml without having to access any of the fields. It seems like for this case, we just need to retrieve the OSFamily, so this looks good to me if we want to generalize how we are retrieving the values from MachineConfig.

File: pkg/support/runsupportbundle_test.go
Comment (danbudris) [inline pkg/support/runsupportbundle_test.go:270]: this test needs to be improved, and actually test that the correct collectors are being called.

File: pkg/support/collectors.go
Comment (jaxesn) [inline pkg/support/collectors.go:128]: Would it make sense to use an array or map here that could be easier to add additional filenames/paths to instead of duplicating the analyzer object over and over?

Comment (danbudris) [reply pkg/support/collectors.go:128]: yeah , we could simplify this. for the deployments, [we're using a simplified struct that has just the variable information in it](https://github.com/aws/eks-anywhere/blob/main/pkg/support/analyzers.go#L19), or we could use a map that's [collector name]collectionPath, and just set everything else the same. The only field that's really variable at this point is timeout, and the smaller collectors won't be harmed by having a larger timeout

File: pkg/support/collectors.go
Comment (danbudris) [reply pkg/support/collectors.go:121]: ya makes sense

File: cmd/eksctl-anywhere/cmd/generatebundleconfig.go
Comment (danbudris) [reply cmd/eksctl-anywhere/cmd/generatebundleconfig.go:84]: yeah good call

File: pkg/support/collectors.go
Comment (danbudris) [reply pkg/support/collectors.go:128]: we can also easily update this after this PR merges

File: pkg/support/collectors.go
Comment (danbudris) [reply pkg/support/collectors.go:121]: this is intended to easily support heterogeneous OS families in the future. I'm using a 'set' so we 

- only end up with a single entry per os family, make it easy to avoid generating redundant configs. 
- allow the 'collectorFactory' to determine what collectors we generate, making it easier to determine shared collectors, group collectors, etc...this is going to increase in complexity once we add CP vs WN specific collectors, heterogeneous clusters with both BR & Ubuntu, etc, so I think keeping all the logic in the factory will be good

that being said, we can get rid of the switch and just check for the presence of the specific OS family in the map with 'ok' or w/e, that's fine by me

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/286#" title="Author self-approved">danbudris</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/support/collectors.go
Comment (danbudris) [reply pkg/support/collectors.go]: ok updated

File: cmd/eksctl-anywhere/cmd/generatebundleconfig.go
Comment (danbudris) [reply cmd/eksctl-anywhere/cmd/generatebundleconfig.go]: done

File: test/framework/flux.go
Comment (danbudris) [reply test/framework/flux.go:234]: the values being passed in, `vsphereMachineConfigs`, is a map of string: pointer to machine config. 

So by using a pointer TO THE POINTER, or basically `**vsphereMachineConfig`, we're referencing the concrete type, which does not implement the newly added `OSFamily()` method of the `MachineConfig` interface, as it has a pointer receiver on the `VsphereMachineConfig`

so here we're just using the given pointer from the array, rather than the concrete type `**vsphereMachineConfig`

I'll add a nil check for safey sake, but I think it's pretty safe without it given the current usage and the non-result of appending nil

**Comment by abhay-krishna** [comment]
@danbudris: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | 328837f5f9e5706762df11fcabaacd5cf3bfcc24 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/286/eks-anywhere-cluster-controller-tooling-presubmit/1441088059838828544) | `/test eks-anywhere-cluster-controller-tooling-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=286). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:danbudris).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by vivek-koppuru — APPROVED** [review]
/lgtm

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/286#" title="Author self-approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/286#pullrequestreview-762420039" title="LGTM">vivek-koppuru</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,vivek-koppuru]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by abhay-krishna** [comment]
@danbudris: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | 328837f5f9e5706762df11fcabaacd5cf3bfcc24 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/286/eks-anywhere-cluster-controller-tooling-presubmit/1441088059838828544) | `/test eks-anywhere-cluster-controller-tooling-presubmit`
eks-anywhere-attribution-files-presubmit | ba98b5b2da0defe76529305a23ad25c26589fa6e | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/286/eks-anywhere-attribution-files-presubmit/1441121542741168128) | `/test eks-anywhere-attribution-files-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=286). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:danbudris).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@danbudris: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | 328837f5f9e5706762df11fcabaacd5cf3bfcc24 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/286/eks-anywhere-cluster-controller-tooling-presubmit/1441088059838828544) | `/test eks-anywhere-cluster-controller-tooling-presubmit`
eks-anywhere-attribution-files-presubmit | ba98b5b2da0defe76529305a23ad25c26589fa6e | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/286/eks-anywhere-attribution-files-presubmit/1441128564689735680) | `/test eks-anywhere-attribution-files-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=286). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:danbudris).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Pull request merged**
